### ISSUE TITLE
Support automatic reload of rotated certificates and private keys

### ISSUE THREAD

**Issue by tsaarni** [keycloak/keycloak#26524]
### Description

Feature request: 

Keycloak should automatically reload certificates and private keys when they change, without having to restart the server. Sometimes this feature is referred to as "hot-reload" or "hitless reload".

### Discussion

https://github.com/keycloak/keycloak/discussions/10654

### Motivation

Currently, Keycloak needs to be manually restarted for example when HTTPS server credentials are renewed. Why is this a problem? Let's Encrypt recommends renewing certificates [every 60 days](https://letsencrypt.org/docs/faq/#what-is-the-lifetime-for-let-s-encrypt-certificates-for-how-long-are-they-valid). A lot shorter renewal period may be used in certain deployments in internal "private" PKI. For example certificates used inside Kubernetes cluster could be rotated within hours or minutes. It becomes inconvenient to setup monitoring for the expiration and to restart Keycloak every time certificate is rotated. For critical and highly available services such as Keycloak it is desirable to support automatic certificate reload after rotation.

### Details

The feature is not limited to HTTPS server certificate. Any client or server certificates and private key that Keycloak uses is updated (causing same service disruption) and could be automatically reloaded. Also CA certificates may be updated, or new ones might need to be configured e.g. as result of adding new a realm with new federated server that needs to be validated with CA certificate.

Previous related issues & work:

* Feature request for Eclipse Vert.x https://github.com/eclipse-vertx/vert.x/issues/3780
* Feature request for Quarkus https://github.com/quarkusio/quarkus/issues/15926
* Proposal for Keystore SPI https://github.com/keycloak/keycloak-community/pull/345
* Feature request for dynamic truststores https://github.com/keycloak/keycloak/issues/13300

**Comment by ahus1** [comment]
Thank you for this suggestion, and also the linking to the various topics. Hot reloading of certificates is also the default on Kubernetes when using things like certmanager.

From the operations perspective, the keystores should be hot-reloadable by default. Look at your code at https://github.com/Nordix/keycloak, you are using the keystore / truststore SPI to enable what you describe here, and that includes the customizing a keystore for Quarkus.

Maybe you can discuss with @vmuzikar and @shawkins this can be simplified if the hot reloading of certificates is the new default. 

They can also guide you to find out what if those changes should be applied to Keycloak, or of some of the changes should go into Quarkus.

**Comment by vmuzikar** [comment]
@tsaarni Thanks for the RFE.

Hot reload seems like a reasonable request. On the other hand, a Keycloak deployment restart is certainly not something very rare. Restart is needed for every config change, certs included. Pods comes and goes. Possibly the focus should be on the startup time, to make the restart as fast as possible.

Another related (possibly duplicated issue): #23771

**Comment by vmuzikar** [comment]
> During a restart, to not lose requests a Keycloak Pod would require a "slow" shutdown as one should possibly first remove it from the loadbalancer (marking it "not ready" in Kubernetes terms), then wait for requests to finish, and only then it should be shut down - something that should be doing, but are currently not AFAIK.

Yes, we're not doing that. And I can see that this might be something we'd want to look into prior to considering hot-reload as deployment restarts are inevitable in any case (if not for config changes, then for often micro releases and image respins due to CVEs).

> Would you, @vmuzikar, be willing to look at a community PR and mark this as "help wanted"? While the sole HTTPS certificate rotation might be a Quarkus thing, rotating the other certificates is IMHO not.

The main issue I see here is the unclear scope of this. Hot-reload is mostly relevant for TLS cert/key which is currently depending on Quarkus. As soon as Quarkus supports this, we can too.

Another potential area is the Keycloak truststore (#13300). IMHO, this might not be very straightforward. With the recent efforts to provide a unified truststore experience, we've introduced a system truststore (#24473). This is now piggybacking on Java system trustore which is not implicitly hot-reloadable, AFAIK. This might depend on each implementation that is using the system trustore to make sure it's hot-reloadable which might introduce some inconsistencies. Some components will see the old trustore, some will automatically load the new. From the Keycloak perspective, we control only the Keycloak Truststore SPI which is only one of the components now using the system truststore. However, we're trying to not put too much emphasis on the Truststore SPI as it was confusing and often unclear where it's used and where it's not used. It's a bit hidden from users now. If we added hot-reload support to that, we'd need to highlight it more to the user once again. Not sure if it's desirable. WDYT?

Another possible area for hot-reaload might be the embedded ISPN keystore/truststore.

**Comment by ecerulm** [comment]
How is this proposal different from  #23771? I mean, if this one supersede the one I created at #23771 then I'm fine with closing #23771 in favor of this one. 

The idea of restarting Keycloak when the certbot renews the certificates, is certainly the base solution and can be easily done with cerbot deploy hooks. The rationale behind #23771 was to have **zero downtime**. Using this solution implies **some downtime every 60 days.**. 

The other solution (apart from the certbot deploy hook) is to use a reverse proxy like Caddy that will renew the certificates itself without downtime. 

I don't know about the other certificates, when I wrote #23771 I was concerned only with the HTTPS  url .

**Comment by vmuzikar** [comment]
@ecerulm As far as I understand, this issue is has larger scope compared to #23771 and concerns any keys/certs in Keycloak in general. #23771 is covering only TLS keys/certs.

**Comment by vmuzikar** [comment]
CC @keycloak/cloud-native

**Comment by tsaarni** [comment]
Also in our use case we are after zero downtime. It is true of course that for software upgrades it is not possible to achieve, but I wish it does not mean it could not be the goal for cases where it is possible to achieve.

To tell the truth, I missed #23771 when submitting this, but the difference is indeed that there are few other types of TLS certs beyond HTTPS server cert. There could be a common solution that could cover (some) of the cases in unified way, so I thought having a "blanket" issue could be appropriate but I'm not against taking a look at them individually either :smile: 

My use cases includes client certificates for certificates for LDAP federation and for identity brokering with external OpenID server over HTTPS. People have asked about rotating trusted certificates as well (dynamically #13300 or just generally by _any_ method that would avoid restart). Besides these certs, there is also SMTP. I know pgjdbc already reloads certs at every new connection and my colleague previously added support for [JGroups](https://www.github.com/belaban/JGroups/pull/692).  Maybe it would be good to have these written down somewhere.

When I worked with this topic previously I made an attempt to introduce hot-reload for HTTPS server cert in Vert.X, but at the end it was preferred to be done at application / upper layers. After that, I looked into alternative of having `Keystore` implementation that internally can hot-reload underlying files, and could be shared as a "black box" between any TLS scenarios which accepts custom `KeyManagers` (like many does). Java TLS implementation does not support hot-reloading `TrustManagers` (why? I added [notes here](https://github.com/tsaarni/reloading-keystore/blob/main/docs/implementation-description.md#why-hot-reload-does-not-work-for-truststore)).

I'm not against solving only Quarkus case individually if that is the preference. I can contribute if we reach agreement there that Quarkus can take this functionality to watch file paths and trigger Vert.X re-configuration, instead of delegating that to application via `HttpServerOptionsCustomizer`.

**Comment by sschu** [comment]
Personally, I would much rather go into the direction of painless restarts like @vmuzikar suggested instead of adding yet another functionality to Keycloak itself.
@ahus1 How would the operator support connection draining? Isn't that something the ingress would have to support? Or does the operator already feature an ingress?

**Comment by ahus1** [comment]
>  Isn't that something the ingress would have to support? 

Yes, and this is what makes it painful to say the least. See below. As it is so very painful, I'd be +1 for hot reloading.

Part of the way is to make the Pod respond to the readiness probe with "not-ok", so that the loadbalancer will direct new connections to other Pods. When a Pod becomes non-ready depends on the number of retries and the interval configured in the readiness probe. And then it takes some more time (1-2 seconds?) for the ingress to reconfigure itself.

After that the next step (But only then! Unfortunately the application in the pod doesn't have a clue when the first step is complete!): If you have TLS passthough, the loadbalancer will only see the connection, not the HTTP requests on it. The connections are usually HTTP keepalive, With HTTP/1.1 you might be able to disable keepalive so that you are able to close a connection after the *next* request of the client. It is even worse with HTTP/2 multiplexing when multiple requests are handled over the connection. Eventually the server must close all connections in the most graceful moment and exit, but only after the loadbalancer has stopped forwarding new requests.

**Comment by sschu** [comment]
But if you do an active operation like rollout restart on k8s, you don't need to detect that the pod is unready. The operation basically tells the load balancer that the pod will be unready and the ingress stops sending more request to the pod.

**Comment by ahus1** [comment]
> But if you do an active operation like rollout restart on k8s, you don't need to detect that the pod is unready.

IMHO this works but unfortunately the the wrong way around: First the pod will be down and no longer answering requests, and then the loadbalancer will know about it and start reconfiguring itself. Kubernetes is a distributed system and everyone is listening for changes on some resources on etcd :-/ happy to be proven wrong.

**Comment by tsaarni** [comment]
In our case, we use TLS also inside the cluster and we'd like to use very short-lived certificates (in hours, rather than 90 days). We have other applications within cluster communicating with Keycloak directly with TLS and also federation towards an LDAP server inside the same cluster.

**Comment by sschu** [comment]
Pod restarts is what Kubernetes might do at any time for numerous reasons. That's why I think they should not cause pain.

**Comment by tsaarni** [comment]
Surely the idea in cloud-native is to have everything disposable, but restarts are still often not painless / without in-service impact, even more so in paths that does not have proxy. Users may wish that common reconfiguration scenarios would need not revert back to restart. In general, I don't think there is anything special in Kubernetes that makes it restart workload any more than any other clustering environment.

**Comment by ahus1** [comment]
IMHO the interest of seamless restarts would be discussed better in a separate issue. Someone interested in such a discussion could search the existing issues, and then either link it here or create a new one.

**Comment by ecerulm** [comment]
> Pod restarts is what Kubernetes might do at any time for numerous reasons

Is  Keycloak project targeting Kubernetes exclusively? Honest question, when I tried Keycloak I was planning for a "conventional" deployment on a linux server, and now I wonder if it's "supported" at all.

**Comment by ahus1** [comment]
> Is Keycloak project targeting Kubernetes exclusively? Honest question, when I tried Keycloak I was planning for a "conventional" deployment on a linux server, and now I wonder if it's "supported" at all.

Yes, plain deployment on a (Linux) server is fully supported, and will be supported in the future. I know that a lot of people use it this way. 

At the same time, if we would optimize Keycloak for seamless and graceful restarts because we think they are happening frequently in a cloud environment, we might ask people on bare metal to use the seamless restart functionality as well once it is there (and not earlier) if that solves the problems about certificates. Please read it as "ask" in the natural sense, we wouldn't be forcing this. And when we ask, we would also listen to reasons when those restart don't make sense outside of cloud environments. 

The effect @vmuzikar described above, that seamless restarts for configuration changes would benefit everyone's situations (regardless of cloud or not) applies as well.

**Comment by kfox1111** [comment]
When integrating with SPIRE, the certificates get rolled quite frequently. A full container restart would be a lot of extra work in that environment.

**Comment by ahus1** [comment]
> When integrating with SPIRE, the certificates get rolled quite frequently.

Can you be more specific on the number / when this happens in your environment?

**Comment by kfox1111** [comment]
The default ttl is 1h. (https://spiffe.io/docs/latest/deploying/spire_server/#server-configuration-file, option default_x509_svid_ttl	)

**Comment by hafiz-qasim** [comment]
we have the same issue that keycloak restart is needed to update certificates for example for LDAP,
Are there any plans to have this enhancement in the near future?
Is it possible to have a REST api to update certificates in keycloak?

**Comment by squaricdot** [comment]
My plan so far was to use Kyverno and check if the secret changed that is held by the cert-manager's Certificate object. When it changes trigger restart of the keycloak deployment/statefulset. That is quite optimistically speaking perhaps. If you have a HA keycloak cluster, the sessions are distributed in infinispan. Graceful restart should be fairly safe right

**Comment by ahus1** [comment]
@squaricdot - Rolling restarts using a stateful set should not loose sessions. Still, those restarted nodes need to warm up (fill their caches, warm up the JVM), so you might see that the first requests they are serving are slow.

**Comment by kfox1111** [comment]
That is not going to work very well for spire....

With spire, each pod gets individual certs issued and pushed to it. The timing for each one may not match exactly all the other pods.

So, it may take multiple rollouts to get all of them at the orchestration level, or would require restarting the individual pods along with some kind of locking to ensure not too many go out at once.

We're talking about doing hacky things to work around doing the right thing IMO. The process should just support refreshing natively.

**Comment by kfox1111** [comment]
Did this get implemented as part of https://github.com/keycloak/keycloak/pull/32715 ?

**Comment by shawkins** [comment]
> Did this get implemented as part of https://github.com/keycloak/keycloak/pull/32715 ?

#32715 implements a subset of this support - only keys and certs that are specified as keycloak https properties.

Any other usage of certs / keys including via the --truststore-paths option do not automatically reload.

> So, it may take multiple rollouts to get all of them at the orchestration level

A rolling restart of the statefulset that @ahus1is mentioning would look like using kubectl rollout restart or just adding a custom annotation to the statefulset pod spec.

The default strategy for the statefulset would be to do this serially - https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#rolling-updates

> or would require restarting the individual pods along with some kind of locking to ensure not too many go out at once.

You can of course individually delete Pods to have them be recreated, and can consider a PodDisruptionBudget if you want protection on how many pods stay in service.

**Comment by kfox1111** [comment]
Ah, ok. How much effort would it be to do the rest?

Yeah, I guess I'm suggesting orchestrated rollout of any kind is just not a good strategy when the cert reload is so dynamic in some systems.

**Comment by shawkins** [comment]
> Ah, ok. How much effort would it be to do the rest?

A decent amount. Each additional cert / key usage to be configured through the Quarkus TLS Registry instead of how it's being directly wired today.

> Yeah, I guess I'm suggesting orchestrated rollout of any kind is just not a good strategy when the cert reload is so dynamic in some systems.

Yes, it's not ideal.

**Comment by dvakarchuk-iqmessenger** [comment]
Thanks for opening and discussing this feature request, and for the pointers to the related issues and the new truststore support.

In addition to hot‑reloading TLS certificates, it would be very helpful if Keycloak could also automatically reload trusted certificates used via the new truststore mechanism.

Right now, truststores are discovered from `conf/truststores` and from any additional locations specified via `--truststore-paths`, but they appear to be read only at startup, so rotating a CA or intermediate certificate still requires a full Keycloak restart to take effect. This is painful in setups where external services rotate their certificates frequently (internal PKI, Kubernetes service meshes, LDAPS endpoints, external IdPs, SMTP, etc.), because it prevents zero‑downtime rotation for outgoing TLS connections.

It would be great if the same approach you are considering for server TLS cert hot reload could also be applied to the truststore: for example, by watching the configured truststore paths and rebuilding/reloading the system truststore (or the TruststoreProvider / Truststore SPI) when files under those paths change, instead of requiring a full server restart.

Also, to make this operationally consistent, it could reuse the same kind of polling mechanism/config style as the existing `https-certificates-reload-period` option (or an analogous `truststore-reload-period` / unified reload-period setting), so operators can control reload frequency the same way for both key material and trust anchors.

If this is considered out of scope for this issue, would it make sense to track “truststore hot reload” as a separate but related enhancement, reusing whatever mechanism you end up implementing here for certificate hot reload?

Thanks again for looking into this and for all the work on improving certificate and truststore handling in Keycloak.

**Comment by stianst** [comment]
If the main issue is BouncyCastle truststore is an issue; then maybe we can deprecate/remove support for that if we have valid alternatives for FIPS now

**Comment by tsaarni** [comment]
I don't think BouncyCastle is the problem here.

I know two ways to do hot-reload:

1. Re-initialize `SSLContext` with a new `KeyStore`/`TrustStore`.
2.  Re-implement `KeyStore` so it updates itself without re-initializing.

Our current HTTPS reload setup (from Quarkus/Vert.x) uses approach (1). It works for both server certs and CA certs.

I don't think we can use (1) for e.g. LDAP because the `SSLContext` handling is buried deep in the JDK. We can use approach (2) (I did this for LDAP federation in an old PR), but it won't work for reloading CA certificates. I've explained why hot-reload does not work for truststore [here](https://github.com/tsaarni/reloading-keystore/blob/main/docs/implementation-description.md#why-hot-reload-does-not-work-for-truststore) (`TrustManager` makes a copy of certificates).

**Comment by shawkins** [comment]
> I don't think BouncyCastle is the problem here.

It is to reuse the quarkus TLS registry logic. https://github.com/quarkusio/quarkus/issues/50304

**Comment by tsaarni** [comment]
> It is to reuse the quarkus TLS registry logic. [quarkusio/quarkus#50304](https://github.com/quarkusio/quarkus/issues/50304)

I'm a little confused :) Is that also related to credential (cert+key) and CA certificate hot-reload?

**Comment by shawkins** [comment]
> I'm a little confused :) Is that also related to credential (cert+key) and CA certificate hot-reload?

Yes, I would like to align things as much as possible to the support that quarkus already provides.

**Comment by tsaarni** [comment]
> Yes, I would like to align things as much as possible to the support that quarkus already provides.

Absolutely! Though I still must admit I did not follow how BC removal or adding PKCS#8 encrypted private key support via Quarkus unblock hot reload where it is not currently supported. But that is ok :)

**Comment by shawkins** [comment]
The quarkus TLS registry supports reloading and encrypted private keys. 

They do allow reloading support via the legacy tls configuration, which we are using for our http interface.

**Comment by shawkins** [comment]
Linking to #49968 as that change coupled with https://github.com/keycloak/keycloak/issues/51680 would provide us an http client that is using reloadable certs.
